File: typescript/problem_32.ts
export function hasArbitrage(rates: number[][]): boolean {
    const n = rates.length;
    const dist = new Array(n).fill(0);
    for (let k = 0; k < n - 1; k++)
        for (let i = 0; i < n; i++)
            for (let j = 0; j < n; j++) {
                if (rates[i][j] === 0) continue;
                const w = -Math.log(rates[i][j]);
                if (dist[i] + w < dist[j]) dist[j] = dist[i] + w;
            }
    for (let i = 0; i < n; i++)
        for (let j = 0; j < n; j++) {
            if (rates[i][j] === 0) continue;
            if (dist[i] + (-Math.log(rates[i][j])) < dist[j]) return true;
        }
    return false;
}

export function problem32(): void {
    console.log("Problem 32:");
    const arb  = [[1,0.9,0.8],[1.2,1,0.9],[1.4,1.2,1]];
    const fair = [[1,2,4],[0.5,1,2],[0.25,0.5,1]];
    if (!hasArbitrage(arb))  throw new Error("assertion failed");
    if ( hasArbitrage(fair)) throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: rust/problem_32.rs
pub fn has_arbitrage(rates: &[Vec<f64>]) -> bool {
    let n = rates.len();
    let mut dist = vec![0.0f64; n];
    for _ in 0..n-1 {
        for i in 0..n {
            for j in 0..n {
                if rates[i][j] == 0.0 { continue; }
                let w = -rates[i][j].ln();
                if dist[i] + w < dist[j] { dist[j] = dist[i] + w; }
            }
        }
    }
    for i in 0..n {
        for j in 0..n {
            if rates[i][j] == 0.0 { continue; }
            let w = -rates[i][j].ln();
            if dist[i] + w < dist[j] { return true; }
        }
    }
    false
}

pub fn problem32() {
    println!("Problem 32:");
    let arb  = vec![vec![1.0,0.9,0.8],vec![1.2,1.0,0.9],vec![1.4,1.2,1.0]];
    let fair = vec![vec![1.0,2.0,4.0],vec![0.5,1.0,2.0],vec![0.25,0.5,1.0]];
    assert!( has_arbitrage(&arb));
    assert!(!has_arbitrage(&fair));
    println!("All Tests Passed");
}
